### DISCUSSION TITLE
OpenTelemetry instrumentation for MediatR

### DISCUSSION THREAD

**Discussion by julealgon** [LuckyPennySoftware/MediatR#762] [Ideas]
Every major framework is now being instrumented with OpenTelemetry (aspnet itself, http, ef, etc). I think it would make sense for MediatR to have its own instrumentation library (or built-in instrumentation) to work with OpenTelemetry standards.

Thoughts? Has this been considered before? 

From a developer perspective, it would be nice to have custom spans for MediatR operations show up in traces, as well as any metrics that would be particular to the library.

**Comment by SilentBlueD666** [comment]
I was just working on this yesterday; I simply created a pipeline behaviour.

Or are you saying out of the box would be better?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Juliano Leal Goncalves ***@***.***>
Sent: Tuesday, July 12, 2022 11:31:25 PM
To: jbogard/MediatR ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jbogard/MediatR] OpenTelemetry instrumentation for MediatR (Discussion #762)


Every major framework is now being instrumented with OpenTelemetry (aspnet itself, http, ef, etc). I think it would make sense for MediatR to have its own instrumentation library (or built-in instrumentation) to work with OpenTelemetry standards.

Thoughts? Has this been considered before?

From a developer perspective, it would be nice to have custom spans for MediatR operations show up in traces, as well as any metrics that would be particular to the library.

**Reply by julealgon** [reply]
> I was just working on this yesterday; I simply created a pipeline behaviour. 

I'm not sure what you mean. Are you saying you instrumented it on your own inside your application?

> Or are you saying out of the box would be better?

This question is also not clear to me. There are basically 2 ways the instrumentation could be provided "out of the box":
1. The main `MediatR` package itself has it built-in
2. There is a separate `OpenTelemetry.Contrib.Instrumentation.MediatR` type of package that adds the instrumentation

Out of the 2 above, I actually don't have a strong opinion on which is the preferred approach (I don't know if 1 is a good practice or not).

Hope this clarifies it.

**Comment by jbogard** [comment]
I dunno, I'm not sure if it would add more noise to your traces. The frameworks you listed are all dealing with I/O, so I'd already have traces around incoming HTTP requests, any data access, any messages consumed/produced, and any outgoing HTTP requests. I don't know what another trace would add.

**Reply by julealgon** [reply]
That's fair. I'm also not 100% sure myself if this would be a good idea or not but I figured I'd ask since I'm working quite heavily on OTel stuff lately.

What I can say from a consumer perspective, is that I've already created tracing decorators for MediatR in the past to integrate with XRay (AWS's custom tracing solution) and that actually helped me getting a better _domain_ view of my operations.

Sometimes, having just the HTTP request information is a bit too "raw", if you know what I mean. MediatR is usually used to model something closer to the domain that has more semantic value involved and such. The additional span in the trace exposing this more semantic information was fairly useful in my scenario.

Also, I want to add that it would be quite nice to have MediatR tracing for worker-type processes (i.e. services that run on a pure console, or a windows service, etc). Those are not necessarily activated by an external event (like an HTTP request or a queue push), so without custom tracing, they would not output anything even with all the existing instrumentation packages added. MediatR could be the tracing starter in these scenarios.

**Reply by jbogard** [reply]
That's fair, but it's pretty easy to add via behaviors now that the Activity/Diagnostics API has been improved.
